crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -0,0 +1,732 @@
+///     def inner():
+///         pass
+///
+///     def inner2():
+///         pass
+/// ```
+///
+/// ## References
+/// - [PEP 8](https://peps.python.org/pep-0008/#blank-lines)
+/// - [Flake 8 rule](https://www.flake8rules.com/rules/E306.html)
+#[violation]
+pub struct BlankLinesBeforeNestedDefinition(pub u32);
+
+impl AlwaysFixableViolation for BlankLinesBeforeNestedDefinition {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let BlankLinesBeforeNestedDefinition(blank_lines) = self;
+        format!("Expected 1 blank line before a nested definition, found {blank_lines}")
+    }
+
+    fn fix_title(&self) -> String {
+        "Add missing blank line".to_string()
+    }
+}
+
+/// Returns `true` if the token is Async, Class or Def
+fn is_top_level_token(token: &Option<Tok>) -> bool {
+    matches!(&token, Some(Tok::Class | Tok::Def))
+}
+
+/// Returns `true` if the token is At, Async, Class or Def
+fn is_top_level_token_or_decorator(token: &Tok) -> bool {
+    matches!(&token, Tok::Class | Tok::Def | Tok::At)
+}
+
+#[derive(Debug)]
+struct LogicalLineInfo {
+    first_token: Tok,
+    first_token_range: TextRange,
+    last_token: Tok,
+    last_token_range: TextRange,
+    is_comment_only: bool,
+    is_docstring: bool,
+    indent_level: usize,
+    blank_lines: u32,
+    preceding_blank_lines: u32,
+    preceding_blank_characters: TextSize,
+}
+
+/// Iterator that processes tokens until a full logical line (or comment line) is "built".
+/// It then returns characteristics of that logical line (see `LogicalLineInfo`).
+struct LinePreprocessor<'a> {
+    tokens: Flatten<Iter<'a, Result<(Tok, TextRange), LexicalError>>>,
+    locator: &'a Locator<'a>,
+    /// Number of previous consecutive blank lines.
+    previous_blank_lines: u32,
+    /// Number of consecutive blank lines.
+    current_blank_lines: u32,
+    /// Number of blank characters in the blank lines (\n vs \r\n for example).
+    current_blank_characters: TextSize,
+}
+
+impl<'a> LinePreprocessor<'a> {
+    fn new(tokens: &'a [LexResult], locator: &'a Locator) -> LinePreprocessor<'a> {
+        LinePreprocessor {
+            tokens: tokens.iter().flatten(),
+            locator,
+            previous_blank_lines: 0,
+            current_blank_lines: 0,
+            current_blank_characters: 0.into(),
+        }
+    }
+}
+
+impl<'a> Iterator for LinePreprocessor<'a> {
+    type Item = LogicalLineInfo;
+
+    fn next(&mut self) -> Option<LogicalLineInfo> {
+        let mut line_is_comment_only = true;
+        let mut is_docstring = true;
+        let mut first_token: Option<Tok> = None;
+        let mut first_token_range: Option<TextRange> = None;
+        let mut last_token: Option<Tok> = None;
+        let mut parens = 0u32;
+
+        while let Some((token, range)) = self.tokens.next() {
+            if matches!(token, Tok::Indent | Tok::Dedent) {
+                continue;
+            }
+
+            if !matches!(token, Tok::Newline) {
```

Comment:
I've done the changes in f3305841b.
Is it faster to do a `==` comparison than a `matches!` ? Or is it for readability reasons ?

Reply:
It's mostly for readability. I find `matches!` rather verbose to read, and negating matches always breaks my flow (because there are two `!`, one because it is a macro, and the other for negating.